src/main/scala/io/iohk/ethereum/crypto/zksnark/BN128.scala

```diff
@@ -172,4 +184,57 @@ object BN128 {
     def isValid: Boolean =
       x.isValid() && y.isValid() && z.isValid()
   }
+
+  case class BN128G1(p: Point[Fp])
+  object BN128G1 {
+
+    /**
+      * Constructs valid element of subgroup `G1`
+      * To be valid element of subgroup, elements needs to be valid point (have valid coordinates in Fp_2 and to be on curve
+      * Bn128 in Fp
+      * @return [[scala.None]] if element is invald group element, [[io.iohk.ethereum.crypto.zksnark.BN128.BN128G1]]
+      */
+    def apply(xx: ByteString, yy: ByteString): Option[BN128G1] = {
```

Comment:
I am not sure if apply returning an Option is the best pattern. Maybe you could use some meaningful name for this method.

Reply:
it is fairly standard smart constructor pattern. Why do you think it is not the best?

Reply:
I am not sure, maybe it is just my personal preference;) When I have a case class I expect that apply returns me an instance of this class.